Lg/rust js type mapping (#180)
dfinity/candid: docs/modules/candid-guide/pages/candid-howto.adoc

For more information about the tool that creates a Web interface from the Candid interface of any canister, see the link:https://github.com/dfinity/candid/tree/master/tools/ui[Candid UI] repository.

== Interact with a service from a Motoko canister

If you are writing a service in Motoko, the Motoko compiler automatically translates the signature of your canister’s top-level `actor` or `actor class` into a Candid description, and the `+dfx build+` command ensures that the service description is properly referenced where it needs to be.

For example, if you want to write a `hello` canister in {proglang} that calls the `counter` canister in {proglang}:

[source, motoko]
....
import Counter "canister:Counter";
import Nat "mo:base/Nat";
actor {
  public func greet() : async Text {
    let result = await Counter.inc(1);
    "The current counter is " # Nat.toText(result)
  };
}
....

The import dependency of counter canister is handled by `+dfx build+`, which ensures that the canister id and the Candid

> description is correctly feed into the {proglang} compiler. The {proglang} compiler then translate the Candid type into

Author: ```suggestion
The {proglang} compiler then translates the Candid type into the appropriate native Motoko type. This translation enables you to call the `inc` method natively—as if it were a Motoko function—even if the `counter` canister is implemented in a different language and even if you do not have the source code for the imported canister.
```

> native Motoko type and enables you to call the `inc` method natively, even if the counter canister is implemented in a

Author: ```suggestion
```

> different language and you do not have the source code. For additional information on the type mapping

Author: ```suggestion
For additional information on the type mapping
```

between Candid and {proglang}, you can consult the link:candid-types{outfilesuffix}[Supported types] reference section.

> The {proglang} compiler and `+dfx build+` also auto-generate the Candid description for the `hello` canister to allow

Author: ```suggestion
The {proglang} compiler and `+dfx build+` command also auto-generate the Candid description for the `hello` canister to allow
```

other canisters or tools to interact with the `hello` canister seamlessly. The generated Candid description is located
in your project build directory at `.dfx/local/canisters/hello/hello.did`.

== Interact with a service from a Rust canister

If you write a canister in Rust, the `+dfx build+` command ensures that the service description is properly referenced where it needs to be. However, you need to write the Candid service description manually following the conventions described in the link:https://github.com/dfinity/candid/blob/master/spec/Candid.md#core-grammar[Candid specification].

For example, if you want to write a `hello` canister in {proglang} that calls the `counter` canister in {proglang}:

[source, rust]
....
use ic_cdk_macros::*;

#[import(canister = "counter")]
struct Counter;

#[update]
async fn greet() -> String {
    let result = Counter::inc(1.into()).await;
    format!("The current counter is {}", result)
}
....

> The import dependency of counter canister is handled by `+dfx build+`, which ensures that the canister id and the Candid

Author: Similar suggestion as above 
```suggestion
When the import macro on the `counter` canister— the `#[import(canister = "counter")]` declaration—is processed by the `+dfx build+` command, the `+dfx build+` command ensures that the `counter` canister identifier and the Candid description are passed to the Rust CDK correctly.
The Rust CDK then translates the Candid type into the appropriate native Rust type. 
This translation enables you to call the `inc` method natively—as if it were a Rust function—even if the `counter` canister is implemented in a different language and even if you do not have the source code for the imported canister.
```

native Rust type and enables you to call the `inc` method natively, even if the counter canister is implemented in a
different language and you do not have the source code. For additional information on the type mapping
between Candid and Rust, you can consult the link:candid-types{outfilesuffix}[Supported types] reference section.

For other canisters and tools to interact with the `hello` canister, you need to manually create a `.did` file:

[source, candid]
....
service : {
    greet : () -> (text);
}
....

There is also an experimental feature to generate Candid service description automatically, see this https://github.com/dfinity/candid/blob/master/rust/candid/tests/types.rs#L99[test case] as an example.

For additional information and libraries to help you create Candid services or canisters in Rust, see the documentation for the https://docs.rs/candid/[Candid crate], link:https://github.com/dfinity/cdk-rs/tree/next/examples[Rust CDK examples] and the link:../rust-guide/rust-intro{outfilesuffix}[Rust tutorials].

== Interact with a service from JavaScript

The https://www.npmjs.com/package/@dfinity/agent[dfinity/agent npm package] includes support for

> importing canisters via Candid.

Author: ```suggestion
importing canisters using Candid.
```

For example, if you want to call the `counter` canister, you can write the following JavaScript program:

[source, javascript]
....
import counter from 'ic:canisters/counter';
import BigNumber from 'bignumber.js';
async () => {
  const result = await counter.inc(new BigNumber(42));
  console.log("The current counter is " + result.toString());
}
....

> The import dependency of counter canister is handled by `+dfx build+` and `webpack`, which ensures that the canister id and the Candid description is correctly feed into the JavaScript program. Behind the scenes, the Candid service description is

Author: similar suggestion
```suggestion
When the import dependency of counter canister is processed by the `+dfx build+` command and the `webpack`configuration, this processing ensures that the canister id and the Candid description are passed to the JavaScript program correctly. Behind the scenes, the Candid service description is
```
